contracts/tokenbridge/ethereum/L1AtomicTokenBridgeCreator.sol

```diff
@@ -0,0 +1,483 @@
+                }(address(standardGatewayTemplate), proxyAdmin, bytes(""))
+            )
+        );
+
+        standardGateway.initialize(
+            getCanonicalL2StandardGatewayAddress(),
+            router,
+            inbox,
+            keccak256(type(ClonableBeaconProxy).creationCode),
+            canonicalL2BeaconProxyFactoryAddress
+        );
+
+        return address(standardGateway);
+    }
+
+    function _deployL1CustomGateway(address proxyAdmin, address router, address inbox, address owner)
+        internal
+        returns (address)
+    {
+        L1CustomGateway customGateway = L1CustomGateway(
+            address(
+                new TransparentUpgradeableProxy{
+                    salt: _getL1Salt(OrbitSalts.L1_CUSTOM_GATEWAY, inbox)
+                }(address(customGatewayTemplate), proxyAdmin, bytes(""))
+            )
+        );
+
+        customGateway.initialize(getCanonicalL2CustomGatewayAddress(), router, inbox, owner);
+
+        return address(customGateway);
+    }
+
+    function _deployL1WethGateway(address proxyAdmin, address router, address inbox) internal returns (address) {
+        L1WethGateway wethGateway = L1WethGateway(
+            payable(
+                address(
+                    new TransparentUpgradeableProxy{
+                        salt: _getL1Salt(OrbitSalts.L1_WETH_GATEWAY, inbox)
+                    }(address(wethGatewayTemplate), proxyAdmin, bytes(""))
+                )
+            )
+        );
+
+        wethGateway.initialize(getCanonicalL2WethGatewayAddress(), router, inbox, l1Weth, getCanonicalL2WethAddress());
+
+        return address(wethGateway);
+    }
+
+    function _deployL2Factory(address inbox, uint256 maxGas, uint256 gasPriceBid) internal returns (uint256) {
+        // encode L2 factory bytecode
+        bytes memory deploymentData = _creationCodeFor(l2TokenBridgeFactoryTemplate.code);
+
+        uint256 maxSubmissionCost = IInbox(inbox).calculateRetryableSubmissionFee(deploymentData.length, 0);
+        uint256 value = maxSubmissionCost + maxGas * gasPriceBid;
+        IInbox(inbox).createRetryableTicket{value: value}(
+            address(0), 0, maxSubmissionCost, msg.sender, msg.sender, maxGas, gasPriceBid, deploymentData
+        );
+
+        return value;
+    }
+
+    function _deployL2Contracts(
+        address l1Router,
+        address l1StandardGateway,
+        address l1CustomGateway,
+        address l1WethGateway,
+        address inbox,
+        uint256 maxGas,
+        uint256 gasPriceBid
+    ) internal returns (uint256) {
+        address l2ProxyAdminOwner = _getRollupOwner(inbox);
+        L2CreationCode memory l2Code = L2CreationCode(
+            _creationCodeFor(l2RouterTemplate.code),
+            _creationCodeFor(l2StandardGatewayTemplate.code),
+            _creationCodeFor(l2CustomGatewayTemplate.code),
+            _creationCodeFor(l2WethGatewayTemplate.code),
+            _creationCodeFor(l2WethTemplate.code)
```

Comment:
can probably put these `_creationCodeFor` on L2 to save some L1 gas

Reply:
Moved to L2, tho we still need `_creationCodeFor ` on L1 to prepare L1 factory deployment data.